### PR TITLE
[BUG-1402] Return HTTP 409 on version confict when creating & updating tenants - added relevant unit test

### LINKED ISSUE #1402: [BUG] Return HTTP 409 on version confict when creating & updating tenants
**Describe the bug**

Hello, I opened https://github.com/opensearch-project/security/issues/1095 earlier, which has recently been fixed (thanks!), but it looks like a HTTP 500 (instead of 409) is being returned for the tenant endpoint as well. This is related to https://github.com/phillbaker/terraform-provider-elasticsearch/issues/206.

When OpenDistro receives parallel requests to create or update tenants (`/_opendistro/_security/api/tenants/{name}`), it returns an HTTP 500 response.

Similar to Elasticsearch itself, these conflict exceptions should return a http 409 instead of a 500 to signal that this is an expect error case and a retry is expected.

**To Reproduce**
1. Issue concurrent PUT requests to the tenant endpoint, e.g. `/_opendistro/_security/api/tenants/tenant_1`, `/_opendistro/_security/api/tenants/tenant_2`
2. See 500

**Expected behavior**
A HTTP 409 instead of HTTP 500.

**Plugins**
Opendistro

**Screenshots**
N/A

**Host/Environment (please complete the following information):**
 - OS: Ubuntu 
 - Version 20.04.2

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by prabhask5**
### Description
Added new TenantsActionApiTest file to put api unit tests for the /api/tenants api call- additionally wrote new unit test to test for parallel PUT /tenant/{name} calls, that they successfully return a 409 call and retry the call.

* Category (Enhancement, New feature, Bug fix, Test fix, Refactoring, Maintenance, Documentation)
Test fix

* Why these changes are required?
There was no test that checked for the behavior mentioned in the linked ticket, so previously we didn't know if the issue still existed or not (it apparently was fixed before), to make sure this doesn't happen again for this particular case, I decided to write a concrete test for it.

* What is the old behavior before changes and new behavior after changes?
Test was added, there's no behavior change?

### Issues Resolved
- #1402 - added unit test to cover bug in this issue

### Testing
Added new tenant action unit test to test parallel tenant api calls.

### Check List
- [x] New functionality includes testing
- [x] New functionality has been documented
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by prabhask5** [comment]
@cwperks This PR was created to address the comment you left under #1402

File: src/test/java/org/opensearch/security/dlic/rest/api/TenantsActionTest.java
Comment (peternied) [inline src/test/java/org/opensearch/security/dlic/rest/api/TenantsActionTest.java:117]: None of the async request logic is needed, after the first `PUT` all future requests will fail with conflict because the tenant has already been created.  You can verify this by setting the threadpool size to 1.

File: src/test/java/org/opensearch/security/dlic/rest/api/TenantsActionTest.java
Comment (peternied) [inline src/test/java/org/opensearch/security/dlic/rest/api/TenantsActionTest.java:20]: Switch to `import org.apache.http.HttpStatus;`, core5 version isn't needed.

Comment (peternied) [reply src/test/java/org/opensearch/security/dlic/rest/api/TenantsActionTest.java:20]: Note; I've filed [1] to have this be caught by automation

- [1] https://github.com/opensearch-project/security/issues/3581

File: src/test/java/org/opensearch/security/dlic/rest/api/TenantsActionTest.java
Comment (peternied) [inline src/test/java/org/opensearch/security/dlic/rest/api/TenantsActionTest.java:47]: These endpoint urls are always the same, no need to have different copies

File: src/test/java/org/opensearch/security/dlic/rest/api/TenantsActionTest.java
Comment (peternied) [inline src/test/java/org/opensearch/security/dlic/rest/api/TenantsActionTest.java:48]: These request bodies aren't verified, I would suggest adding a get on the tenant and then verify the expected description matches.

Comment (peternied) [reply src/test/java/org/opensearch/security/dlic/rest/api/TenantsActionTest.java:48]: Hint; use `assertThat(response.findValueInJson("description"), equalTo("create tenant 1"))` to easily and verbosely make sure the result matches.

File: src/test/java/org/opensearch/security/dlic/rest/api/TenantsActionTest.java
Comment (peternied) [inline src/test/java/org/opensearch/security/dlic/rest/api/TenantsActionTest.java:66]: Further tests are unneeded, so long as the 1 PUT is accepted with CREATED, and the 2nd PUT is rejected with CONFLICT this is working as indented.

File: src/test/java/org/opensearch/security/dlic/rest/api/TenantsActionTest.java
Comment (peternied) [inline src/test/java/org/opensearch/security/dlic/rest/api/TenantsActionTest.java:1]: This test should be added to `TenantInfoActionTest.java`, not be in its own file

**Review by peternied — CHANGES_REQUESTED** [review]

**Comment by peternied** [comment]
@prabhask5 Thanks for the contribution of the test - much appreciated! I've left feedback please address and let me know and I'll take another look.

File: src/test/java/org/opensearch/security/dlic/rest/api/TenantsActionTest.java
Comment (cwperks) [reply src/test/java/org/opensearch/security/dlic/rest/api/TenantsActionTest.java:20]: @peternied Should `org.apache.http.HttpStatus` be used in the integrationTest package to make backporting to 2.x easier? The main branch has already upgraded to org.apache.HttpComponents v5 and v4 is used in 2.x. `main` still has a dependency on v4 because its needed by OpenSAML. See details in https://github.com/opensearch-project/security/issues/2932

Comment (cwperks) [reply src/test/java/org/opensearch/security/dlic/rest/api/TenantsActionTest.java:20]: Let's update the import to `org.apache.http.HttpStatus` which will ensure that the auto backport to 2.x will not have merge conflicts around the package name for these classes.

File: src/test/java/org/opensearch/security/dlic/rest/api/TenantsActionTest.java
Comment (prabhask5) [reply src/test/java/org/opensearch/security/dlic/rest/api/TenantsActionTest.java:117]: Do you mean even if they're not synchronous? This is interesting because when running `curl -X PUT https://localhost:9200/_plugins/_security/api/tenants/tenant_1 -u 'admin:admin' --insecure -d '{"description": "create tenant 1"}' -H 'Content-Type: application/json'` in my terminal twice in a row I get 20X responses, with the first one giving a "created" response, and the second one giving a "updated" response.

Comment (peternied) [reply src/test/java/org/opensearch/security/dlic/rest/api/TenantsActionTest.java:117]: To be honest, I'm not sure - I just changed the setting associated with this line to `1` and noticed that nothing failed - which would indicate the tests aren't verifying something correctly if you'd expect them to fail.  Its possible there is a bug in that isn't causing the failure state to be noticed in the code, maybe the issue is because cause of how that validate loop works:

```
                case HttpStatus.SC_OK:
                    break;
```

However, in general test around 'racing' for a conflict at an API level are going to be flaky typically, so I would generally discourage test of this nature unless there is a strongly deterministic way to get a reproduction, such as being able to control the timing across threads, or use C/ETags to ensure processing order.

**Comment by prabhask5** [comment]
@peternied Alright I refactored the added test to test only two parallel requests and I'm getting a consistent one 409 and one 201 codes, which should be good. The only thing that changes from test run to test run is which request is the "conflict" one, but I think I account for this when checking the description with the get request. Let me know if you need me to make any other changes! TY!

File: src/test/java/org/opensearch/security/dlic/rest/api/TenantInfoActionTest.java
Comment (DarshitChanpura) [inline src/test/java/org/opensearch/security/dlic/rest/api/TenantInfoActionTest.java:138]: why need a final declaration here?

**Review by DarshitChanpura — COMMENTED** [review]
Thank you for adding this test @prabhask5.

File: src/test/java/org/opensearch/security/dlic/rest/api/TenantInfoActionTest.java
Comment (peternied) [reply src/test/java/org/opensearch/security/dlic/rest/api/TenantInfoActionTest.java:15]: This was a suggestion, see the context https://github.com/opensearch-project/security/pull/3580#discussion_r1366912993  What do you think?

File: src/test/java/org/opensearch/security/dlic/rest/api/TenantInfoActionTest.java
Comment (cwperks) [inline src/test/java/org/opensearch/security/dlic/rest/api/TenantInfoActionTest.java:115]: wdyt about moving the request and assertion from L124-126 to here to avoid having to keep track of the index of the request with the SC_CREATED response?

File: src/test/java/org/opensearch/security/dlic/rest/api/TenantInfoActionTest.java
Comment (peternied) [inline src/test/java/org/opensearch/security/dlic/rest/api/TenantInfoActionTest.java:122]: Look into using streams and groupings to solve this problem, you can condense this code into ~4 lines which would also be much more readable.

File: src/test/java/org/opensearch/security/dlic/rest/api/TenantInfoActionTest.java
Comment (scrawfor99) [inline src/test/java/org/opensearch/security/dlic/rest/api/TenantInfoActionTest.java:98]: Please rewrite this test case. See slack for context.

**Review by scrawfor99 — CHANGES_REQUESTED** [review]

**Comment by peternied** [comment]
Async tests are on my mind today, see this PR [1] for an example of how I've implemented async tests + validation.  Those utilities might be useful, feel free to merge my branch onto this change if it helps your development.

- [1] https://github.com/opensearch-project/security/pull/3601

File: src/test/java/org/opensearch/security/dlic/rest/api/TenantInfoActionTest.java
Comment (prabhask5) [reply src/test/java/org/opensearch/security/dlic/rest/api/TenantInfoActionTest.java:98]: Is this about the generative AI tools? This is confusing because I definitely did not use any AI tools to write this code. I did use code from UserApiTest.java for  the async request test helper functions, but I think that's fine to do so (please let me know if it's not). I just want to make it clear that I did not use any generative AI tools and will continue to not do so going forward. Thanks!

File: src/test/java/org/opensearch/security/dlic/rest/api/TenantInfoActionTest.java
Comment (prabhask5) [reply src/test/java/org/opensearch/security/dlic/rest/api/TenantInfoActionTest.java:138]: I previously got a weird java warning asking me to use that, but it looks like it's not required with the current version of this function, resolved!

File: src/test/java/org/opensearch/security/dlic/rest/api/TenantInfoActionTest.java
Comment (scrawfor99) [reply src/test/java/org/opensearch/security/dlic/rest/api/TenantInfoActionTest.java:98]: Hi @prabhask5, thanks for the clarification. I will remove the block on the PR.

**Comment by prabhask5** [comment]
I restructured the test using streams and the AsyncActions.java file from #3601 - let me know of other changes, ty!

**Comment by DarshitChanpura** [comment]
We can wait until #3601 merge to ensure any conflicts are resolved.

**Review by DarshitChanpura — DISMISSED** [review]
Ty for adding the test @prabhask5.

**Comment by peternied** [comment]
@prabhask5 please take a look at the test failures, I'm going to hold off on another review until those are passing. Please let us know if you need help with an issue.

**Comment by davidlago** [comment]
@prabhask5 if you are not ready to continue work on this PR, we can go ahead and close it, and re-open at a later time. Please let us know!

**Comment by prabhask5** [comment]
> @prabhask5 please take a look at the test failures, I'm going to hold off on another review until those are passing. Please let us know if you need help with an issue.

@peternied The errors that I was getting in the test were related to java not being able to find import org.opensearch.test.framework.AsyncActions; when it does exist (which is used in your PR as well), I merged main into this PR to try to solve this error, but if it does not fix it, do you have any fixes going forward?

**Comment by peternied** [comment]
@prabhask5 Checkout this guide [1] on rebasing it might help you get unblocked.

- [1] https://akrabat.com/the-beginners-guide-to-rebasing-your-pr/

**Comment by prabhask5** [comment]
@peternied I think I rebased so feel free to unblock the tests. TY!

**Comment by prabhask5** [comment]
I'm going to make a new PR to see if that fixes things.

**Comment by peternied** [comment]
@prabhask5 Could you link the new PR?

**Comment by prabhask5** [comment]
> @prabhask5 Could you link the new PR?

Here: #3673

**Pull request closed without merging**
